### PR TITLE
Use Engine in .createVis

### PR DESCRIPTION
**Pull request by IagoLast**
#1819 
---

### CODE REVIEW

File: src/engine.js [hunk 1]
```diff
@@ -116,15 +118,16 @@ Engine.prototype.off = function off (event, callback) {
  * Once the response has arrived trigger a 'reload-succes' or 'reload-error' event.
  * 
  * @param {string} sourceId - The sourceId triggering the reload event. This is usefull to prevent uneeded requests and save data.
- * @param {string} forceFetch - ????
+ * @param {boolean} forceFetch - ????
```
Comment (alonsogarciapablo) [inline src/engine.js]: Forces dataviews to fetch data from server after a reload

File: src/engine.js [hunk 2]
```diff
@@ -147,6 +150,19 @@ Engine.prototype.addLayer = function addLayer (layer) {
 
 /**
  * 
+ * Remove a layer from the engine layersCollection
+ * 
+ * @param {layer} layer - A new layer to be removed from the engine.
+ * 
+ * @public
+ * @api
```
Comment (alonsogarciapablo) [inline src/engine.js]: I still not see why we're using `@api` here...

File: src/vis/vis.js [hunk 3]
```diff
@@ -132,13 +115,16 @@ var VisModel = Backbone.Model.extend({
       templateName: vizjson.datasource.template_name
     };
 
-    var windshaftClient = new WindshaftClient(windshaftSettings);
+    this._createEngine(windshaftSettings);
```
Comment (alonsogarciapablo) [inline src/vis/vis.js]: I'd leave the variable declaration out of `_createEngine`:
```
this._engine = this._createEngine(windshaftSettings);
this._layersCollection = this._engine._layersCollection;
this._dataviewsCollection = this._engine._dataviewsCollection;
```

File: src/engine.js [hunk 4]
```diff
@@ -177,7 +207,7 @@ Engine.prototype._onReloadSuccess = function _onReloadSuccess (serverResponse, s
 Engine.prototype._onReloadError = function _onReloadError (serverResponse) {
   var errors = serverResponse;
   this._modelUpdater.setErrors(errors);
```
Comment (alonsogarciapablo) [inline src/engine.js]: @ivan `this._modelUpdater.setErrors` expects instances of [WindshaftError](https://github.com/CartoDB/cartodb.js/blob/1819-use-engine/src/windshaft/error.js).

This is how it was done before: https://github.com/CartoDB/cartodb.js/blob/v4/src/windshaft/map-base.js#L59-L60

We probably need a test for this... 😉

File: src/engine.js (hunk 2 above)
Comment (IagoLast) [reply src/engine.js]: In my mind we are exposing engine so we need a way to add/remove layers. As you know we have a discussion to decide the final public api 😉 so all this @api tags chill probably be removed

**Comment by ivanmalagon** [comment]
@alonsogarciapablo 
I've done the changes to parse properly WindshaftErrors. I've moved that parse code to its own module and added specs to assert the functionality. Take a look, please.

File: src/vis/vis.js [hunk 5]
```diff
@@ -298,9 +298,9 @@ var VisModel = Backbone.Model.extend({
       options.success && options.success();
     }.bind(this);
 
-    var onError = function (error) {
+    var onError = function (errors) {
       this._engine.off(Engine.Events.RELOAD_ERROR, onError);
-      this.setError(error);
+      this.setError(errors);
```
Comment (alonsogarciapablo) [inline src/vis/vis.js]: We're only setting 1 error before. Perhaps we can get the first error from `errors`: `this.setError(errors[0])`

**Comment by alonsogarciapablo** [comment]
@ivan 👏 👏 👏  Thanks!

**Comment by ivan** [comment]
Please stop typing @ivan, it subscribes me and sends me an email

**Comment by alonsogarciapablo** [comment]
I'm sorry @ivan! Have a great day!

**Pull request merged**